addons/im_livechat/models/mail_channel.py

```diff
@@ -191,3 +194,134 @@ def _email_livechat_transcript(self, email):
+            raise ValidationError(_('The new step sequence must be higher than the old one.'))
+
+        # We change the current step to the new step
+        self.livechat_chatbot_current_step_id = step_id
+
+        human_operator = False
+        if step_id.step_type == 'forward_operator':
+            human_operator = self.livechat_channel_id._get_random_operator() if self.livechat_channel_id else False
+            if human_operator:
+                self.with_user(self.env.ref('base.user_root')).add_members(
+                    human_operator.partner_id.ids,
+                    open_chat_window=True,
+                    post_joined_message=False)
+
+                if step_id.message:
+                    # first post the message of the step (if we have one)
+                    posted_message = self._chatbot_post_message(step_id.chatbot_id, step_id.message)
+                else:
+                    posted_message = False
+
+                # then post a small custom 'Operator has joined' notification
+                self._chatbot_post_message(
+                    step_id.chatbot_id,
+                    '<div class="o_mail_notification">%s</div>' % _('%s has joined', human_operator.partner_id.name),
+                    text_to_html=False)
+            else:
+                # don't post anything as the bot in case we did not find an operator
+                # the script will continue normally
+                posted_message = False
+        else:
+            posted_message = self._chatbot_post_message(step_id.chatbot_id, step_id.message)
+
+        if human_operator:
+            self._broadcast(human_operator.partner_id.ids)
+            self.channel_pin(self.uuid, pinned=True)
+
+        return posted_message
+
+    def _chatbot_validate_email(self, email_address, chatbot):
+        email_normalized = email_normalize(email_address)
+
+        posted_message = False
+        if not email_normalized:
+            posted_message = self.with_context(mail_create_nosubscribe=True).message_post(
+                author_id=chatbot.operator_partner_id.id,
+                body=_('This does not look like a valid email, can you please try again?'),
+                message_type='comment',
+                subtype_xmlid='mail.mt_comment',
+            )
+
+        return bool(email_normalized), posted_message
+
+    def _message_post_after_hook(self, message, msg_vals):
+        """
+        This method is called just before _notify_thread() method which is calling the _message_format()
+        method. We need a 'chatbot.mail.message' record before it happens to correctly display the message.
+        It's created only if the author is Odoobot and if the mail channel is linked to a chatbot step.
+        """
+        if self.livechat_chatbot_current_step_id:
+            self.env['im_livechat.chatbot.mail.message'].sudo().create({
```

Comment:
Not sure to clearly understand when chatbot messages are created, sometimes as preparation job, sometimes in post-hook ...

Reply:
As mentioned in an earlier comment, these are now only created here (yay).